### PR TITLE
POC Support `date_bin` on timestamps with timezone

### PR DESCRIPTION
**Pull request by appletreeisyellow**
This PR is a POC to support `date_bin` on timestamps with timezone. There are two things covered in this PR:

1. adds a UDF function `to_local_time()`:
    - only accept 1 input in type `Timestamp(..., *)`
    - return in type `Timestamp(..., None)`
    - for example:
   
```sql
> select to_local_time('2024-04-01T00:00:20Z'::timestamp AT TIME ZONE 'Europe/Brussels');
+---------------------------------------------+
| to_local_time(Utf8("2024-04-01T00:00:20Z")) |
+---------------------------------------------+
| 2024-04-01T00:00:20                         |
+---------------------------------------------+
1 row(s) fetched.
Elapsed 0.010 seconds.

> select to_local_time('2024-04-01T00:00:20'::timestamp AT TIME ZONE 'Europe/Brussels');
+--------------------------------------------+
| to_local_time(Utf8("2024-04-01T00:00:20")) |
+--------------------------------------------+
| 2024-04-01T00:00:20                        |
+--------------------------------------------+
1 row(s) fetched.
Elapsed 0.008 seconds.

> select
  time,
  arrow_typeof(time) as type,
  to_local_time(time) as to_local_time,
  arrow_typeof(to_local_time(time)) as to_local_time_type
from (
  select '2024-04-01T00:00:20Z'::timestamp AT TIME ZONE 'Europe/Brussels' as time
);
+---------------------------+------------------------------------------------+---------------------+-----------------------------+
| time                      | type                                           | to_local_time       | to_local_time_type          |
+---------------------------+------------------------------------------------+---------------------+-----------------------------+
| 2024-04-01T00:00:20+02:00 | Timestamp(Nanosecond, Some("Europe/Brussels")) | 2024-04-01T00:00:20 | Timestamp(Nanosecond, None) |
+---------------------------+------------------------------------------------+---------------------+-----------------------------+
1 row(s) fetched.
Elapsed 0.017 seconds.
```

Combine `to_local_time()` with `date_bin()` will look like:

```sql
> select date_bin(interval '1 day', to_local_time('2024-04-01T00:00:20Z'::timestamp AT TIME ZONE 'Europe/Brussels'));
+----------------------------------------------------------------------------------------------------+
| date_bin(IntervalMonthDayNano("18446744073709551616"),to_local_time(Utf8("2024-04-01T00:00:20Z"))) |
+----------------------------------------------------------------------------------------------------+
| 2024-04-01T00:00:00                                                                                |
+----------------------------------------------------------------------------------------------------+


> select date_bin(interval '1 day', to_local_time('2024-04-01T00:00:20Z'::timestamp AT TIME ZONE 'Europe/Brussels')) AT TIME ZONE 'Europe/Brussels';
+----------------------------------------------------------------------------------------------------+
| date_bin(IntervalMonthDayNano("18446744073709551616"),to_local_time(Utf8("2024-04-01T00:00:20Z"))) |
+----------------------------------------------------------------------------------------------------+
| 2024-04-01T00:00:00+02:00                                                                          |
+----------------------------------------------------------------------------------------------------+
```

2. Support `date_bin` on timestamps with timezone -- the original date_bin works correctly. the confusion happens at the display of the timestamp when there is a timezone. Even though the underlying time, the UTC time is correct, but when there is a timezone, the display automatically adjust the time, which result the display time to be wrong. Adjusting the underlaying time can make the final display time to our expected result.

### CODE REVIEW

File: datafusion/functions/src/datetime/date_bin.rs [hunk 1]
```diff
@@ -702,11 +791,11 @@ mod tests {
...
-                    "2020-09-08T00:00:00+05",
-                    "2020-09-08T00:00:00+05",
-                    "2020-09-08T00:00:00+05",
-                    "2020-09-08T00:00:00+05",
-                    "2020-09-08T00:00:00+05",
+                    "2020-09-07T19:00:00+05:00",
+                    "2020-09-07T19:00:00+05:00",
+                    "2020-09-07T19:00:00+05:00",
+                    "2020-09-07T19:00:00+05:00",
+                    "2020-09-07T19:00:00+05:00",
```
Comment (appletreeisyellow) [inline datafusion/functions/src/datetime/date_bin.rs:798]: This is not right.. this is a change in behavior

**Comment by appletreeisyellow** [comment]
With the implementation in https://github.com/appletreeisyellow/datafusion/pull/9/commits/54f1bdc5e83044d61b4eb278e22559d01140ebb3, there is a change in behavior https://github.com/appletreeisyellow/datafusion/pull/9#discussion_r1662931294, which is not right. 

Then I looked into Postgres to see how it works. This is what I found:
1. psql `date_bin()` displays the result either with timezone in UTC or without timezone. They don't display it in a different timezone, or maybe I'm missing something

```sql
postgres =# SELECT
  date_bin('1 d', '2021-10-31T01:00:00', '1970-01-01T00:00:00'),
  date_bin('1 d', '2021-10-31T01:00:00'::timestamp, '1970-01-01T00:00:00'),
  date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00'),
  date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00'),
  date_bin('1 d', '2021-10-31T01:00:00', '1970-01-01T00:00:00') AT TIME ZONE 'Europe/Brussels',
  date_bin('1 d', '2021-10-31T01:00:00', '1970-01-01T00:00:00') AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels'
;
        date_bin        |      date_bin       |      date_bin       |        date_bin        |      timezone       |        timezone
------------------------+---------------------+---------------------+------------------------+---------------------+------------------------
 2021-10-31 00:00:00+00 | 2021-10-31 00:00:00 | 2021-10-31 00:00:00 | 2021-10-30 00:00:00+00 | 2021-10-31 02:00:00 | 2021-10-30 22:00:00+00
(1 row)
```

This is very different from what DataFusion is currently doing, with cargo build from the [current main branch](https://github.com/apache/datafusion/tree/a4796fa07893cc4db6b6c0f92dab85d2720a52c4):

```sql
> SELECT
  date_bin(interval '1 day', '2021-10-31T01:00:00', '1970-01-01T00:00:00') AS date_bin_1,
  date_bin(interval '1 day', '2021-10-31T01:00:00'::timestamp, '1970-01-01T00:00:00') AS date_bin_2,
  date_bin(interval '1 day', '2021-10-31T01:00:00' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00') AS date_bin_3,
  date_bin(interval '1 day', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00') AS date_bin_4,
  date_bin(interval '1 day', '2021-10-31T01:00:00', '1970-01-01T00:00:00') AT TIME ZONE 'Europe/Brussels' AS date_bin_5,
  date_bin(interval '1 day', '2021-10-31T01:00:00', '1970-01-01T00:00:00') AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels' AS date_bin_6
;
+---------------------+---------------------+---------------------------+---------------------------+---------------------------+---------------------------+
| date_bin_1          | date_bin_2          | date_bin_3                | date_bin_4                | date_bin_5                | date_bin_6                |
+---------------------+---------------------+---------------------------+---------------------------+---------------------------+---------------------------+
| 2021-10-31T00:00:00 | 2021-10-31T00:00:00 | 2021-10-30T02:00:00+02:00 | 2021-10-31T02:00:00+02:00 | 2021-10-31T00:00:00+02:00 | 2021-10-31T02:00:00+02:00 |
+---------------------+---------------------+---------------------------+---------------------------+---------------------------+---------------------------+
1 row(s) fetched.
Elapsed 0.031 seconds.
```

2. Different number of timestamp casting will result in a different date type in terms of timezone:
    - **Odd** number of timestamp casting --> timestamp **without** time zone
    - **Even** number of timestamp casting --> timestamp **with** time zone
    - This postgres behavior is mentioned in https://github.com/apache/datafusion/issues/10368#issuecomment-2111684373,  https://github.com/apache/datafusion/issues/10602#issuecomment-2122811703, and https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT

```sql
postgres =# SELECT
  pg_typeof('2024-03-31T00:30:00') AS type_unknown,
  pg_typeof('2024-03-31T00:30:00' AT TIME ZONE 'Europe/Brussels') AS type_wo_tz, -- 1 casting
  pg_typeof('2024-03-31T00:30:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels') AS type_w_tz, -- 2 casting
  pg_typeof('2024-03-31T00:30:00'::timestamp AT TIME ZONE 'UTC') AS type_w_tz, -- 2 casting
  pg_typeof('2024-03-31T00:30:00'::timestamp AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels') AS type_wo_tz -- 3 casting
;
 type_unknown |         type_wo_tz          |        type_w_tz         |        type_w_tz         |         type_wo_tz
--------------+-----------------------------+--------------------------+--------------------------+-----------------------------
 unknown      | timestamp without time zone | timestamp with time zone | timestamp with time zone | timestamp without time zone
(1 row)
```

**Comment by appletreeisyellow** [comment]
My proposal is:
- For short term fix, use UDF function `to_local_time()` combined with the current implementation of `date_bin()` to achieve the desired result
- For long term fix, we might need to be clear on the timestamp representation in arrow-rs first before we can moving forward: https://github.com/apache/arrow-rs/issues/5827

**Comment by alamb** [comment]
> My proposal is:
> 
> * For short term fix, use UDF function `to_local_time()` combined with the current implementation of `date_bin()` to achieve the desired result

This makes sense to me

> * For long term fix, we might need to be clear on the timestamp representation in arrow-rs first before we can moving forward: [Propose change timestamp casting with timezone to without timezones (also parsing of timestamps without timezones) apache/arrow-rs#5827](https://github.com/apache/arrow-rs/issues/5827)

If we made the change proposed in https://github.com/apache/arrow-rs/issues/5827, would that "solve" all these issues? 

If so, how would someone implement timezone aware date binning?  
Would it be to do like what postgres does and cast back to timestamp without timezone?

**Comment by appletreeisyellow** [comment]
> If so, how would someone implement timezone aware date binning?
Would it be to do like what postgres does and cast back to timestamp without timezone?

I'm not sure. I think we need more discussion to define the behavior.

This is how postgres behaves:
- If the input timestamp has timezone, `date_bin` result will be in the timezone of the system, instead of the input timezone! postgres `date_bin` aware the system timezone, but not the input timezone! 
    - Currently, DataFusion `date_bin` returns the same timezone as the input timezone
    - Do we want to match the behavior with postgres?
- If the input timestamp doesn't have timezone, `date_bin` result will not have timezone

For example, in postgres:

```sql
postgres =# select
  pg_typeof('2021-10-31T01:00:00'::timestamp) as input_datatype,
  date_bin('1 d', '2021-10-31T01:00:00'::timestamp, '1970-01-01T00:00:00'),
  pg_typeof(date_bin('1 d', '2021-10-31T01:00:00'::timestamp, '1970-01-01T00:00:00')) as result_datetype;
       input_datatype        |      date_bin       |       result_datetype
-----------------------------+---------------------+-----------------------------
 timestamp without time zone | 2021-10-31 00:00:00 | timestamp without time zone
(1 row)

postgres=# show timezone;
    TimeZone
-----------------
 America/Chicago
(1 row)

postgres=# select
  pg_typeof('2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels') as input_datatype,
  date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00'),
  pg_typeof(date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00')) as result_datetype;
      input_datatype      |        date_bin        |     result_datetype
--------------------------+------------------------+--------------------------
 timestamp with time zone | 2021-10-30 01:00:00-05 | timestamp with time zone
                                            -- -05 is the system timezone,  it is 01:00 instead of 00:00 because the origin time is '1970-01-01T00:00:00'

# if cast the origin to 'Europe/Brussels' timezone, the result will be at 00:00
postgres=# select date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels');
        date_bin
------------------------
 2021-10-31 00:00:00-05
(1 row)

postgres=# set time zone 'UTC';
SET

postgres=# show timezone;
 TimeZone
----------
 UTC
(1 row)

postgres=# select
  pg_typeof('2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels') as input_datatype,
  date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00'),
  pg_typeof(date_bin('1 d', '2021-10-31T01:00:00' AT TIME ZONE 'UTC' AT TIME ZONE 'Europe/Brussels', '1970-01-01T00:00:00')) as result_datetype;
      input_datatype      |        date_bin        |     result_datetype
--------------------------+------------------------+--------------------------
 timestamp with time zone | 2021-10-30 00:00:00+00 | timestamp with time zone
                                            -- -01 is the system timezone, instead of the input timezone
(1 row)
```

**Comment by appletreeisyellow** [comment]
> My proposal is:
> 
> * For short term fix, use UDF function `to_local_time()` combined with the current implementation of `date_bin()` to achieve the desired result
> * For long term fix, we might need to be clear on the timestamp representation in arrow-rs first before we can moving forward: [Propose change timestamp casting with timezone to without timezones (also parsing of timestamps without timezones) apache/arrow-rs#5827](https://github.com/apache/arrow-rs/issues/5827)

Here is the short term fix PR: https://github.com/apache/datafusion/pull/11347

For long term fix, there are still discussions going on in https://github.com/apache/arrow-rs/issues/5827 which is the first step of the long term fix. 

At this point, I conclude the POC is done, so closing this draft PR.

**Pull request closed without merging**